man/qdepends.1
.\" generated by mkman.py, please do NOT edit!
.TH qdepends "1" "Jan 2024" "Gentoo Foundation" "qdepends"
.SH NAME
qdepends \- show dependency info
.SH SYNOPSIS
.B qdepends
\fI[opts] <pkgname>\fR
.SH DESCRIPTION
The
.I qdepends
applet has two different modes, forward and reverse dependency querying.
Default operation is forward mode, answering the queries "what does
package X depend on", while reverse mode answers "what packages depend
on X".  Both modes can be further specified into build (DEPEND), run
(RDEPEND), post (PDEPEND), EAPI7's build (BDEPEND) and EAPI8's
install-time (IDEPEND) dependencies.
.P
By default, \fIqdepends\fR returns the unique set of atoms that match
all dependency variables.  To split this out per variable, use \fB-v\fR
option.
.P
Currently,
.I qdepends
defaults to query installed packages.  To query packages from ebuild
trees, use the \fB-t\fR option.
.P
If there is no answer to your query (i.e. you've asked for a package
that is not installed, or a version that does not match), then you will
get empty output.
.P
After version \fB0.74\fR of portage-utils, \fIqdepends\fR was changed
considerably to be more consistent and more advanced.  Most notably,
this has changed default output for reverse mode (\fB-Q\fR).  Instead of
just displaying the package name, like for forward mode, all
dependencies for the package are shown.  When colours are enabled, the
matched atom is highlighted in the list.  In addition to just querying
DEPEND, the default mode changed to query all DEPEND-variables, and
return the unique atoms found in them.  Automatic regular expression
match support was removed.  The \fB-N\fR option was removed, as the same
effect can be achieved via the new \fB-F\fR option or \fB-q\fR option.
The \fB-f\fR option was renamed to \fB-S\fR.
.SH OPTIONS
.TP
\fB\-d\fR, \fB\-\-depend\fR
Show DEPEND info.
.TP
\fB\-r\fR, \fB\-\-rdepend\fR
Show RDEPEND info.
.TP
\fB\-p\fR, \fB\-\-pdepend\fR
Show PDEPEND info.
.TP
\fB\-b\fR, \fB\-\-bdepend\fR
Show BDEPEND info.
.TP
\fB\-I\fR, \fB\-\-idepend\fR
Show IDEPEND info.
.TP
\fB\-Q\fR, \fB\-\-query\fR
Query reverse deps.  This basically reverses the search to any
package that references \fI<arg>\fR in DEPEND, RDEPEND, PDEPEND or BDEPEND.
This can be useful to find consumers of a given package, e.g.\ to
search for packages that have \fIlogwatch\fR in their DEPEND.  Note
that using versions or range specifiers may yield odd results since
dependency strings often have ranges themselves.  For installed
packages, SLOT and SUBSLOTs are available, thus SLOT and antiSLOT
queries are possible.  See \fBqatom\fR(1).
.TP
\fB\-i\fR, \fB\-\-installed\fR
Search installed packages using VDB.
.TP
\fB\-t\fR, \fB\-\-tree\fR
Search available ebuilds in the tree.
.TP
\fB\-U\fR, \fB\-\-use\fR
Apply profile USE-flags to conditional deps.
.TP
\fB\-F\fR \fI<arg>\fR, \fB\-\-format\fR \fI<arg>\fR
Pretty-print DEPEND declaration to be used in an ebuild.  This
option initiates a very different mode of operation.  Instead of
printing searching through packages, it constructs a multi-line
statement with shell syntax, to be used in an ebuild.  Each
argument is turned into a separate DEPEND variable declaration.  You
need to quote dependency strings in order for them to be printed as
a single dependency declaration.  When used with the \fB\-q\fR
option, only the pretty-printed dependency declaration is printed,
e.g.\ the DEPEND= part is skipped.
.TP
\fB\-S\fR, \fB\-\-pretty\fR
Pretty format specified depend strings.
.TP
\fB\-R\fR, \fB\-\-resolve\fR
Resolve found dependencies to package versions.
.TP
\fB\-\-root\fR \fI<arg>\fR
Set the ROOT env var.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Pretty-print atoms output, per DEPEND-variable.  This is much like
the \fB-f\fR option, but uses as input the dependency info from the
matches.  When used with \fB-t\fR, the atom highlighting matches
what is used by the installed package, if available.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Suppress DEPEND= output for \fB\-f\fR.  Only print the matching atom for \fB\-Q\fR.  When given two or more times, suppresses the matching atom for \fB\-Q\fR, e.g.\ producing just a list of packages.
.TP
\fB\-C\fR, \fB\-\-nocolor\fR
Don't output color.
.TP
\fB\-\-color\fR
Force color in output.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print this help and exit.
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit.
.SH "EXAMPLES"
For finding out what a particular package depends on for building, you could do:
.nf
	$ \fIqdepends -d nano\fR
	app-editors/nano-2.3.2: >=sys-libs/ncurses-5.9-r1[unicode] sys-apps/file ...
.fi
This tells us that we have \fIapp-editors/nano-2.3.2\fR installed and it depends
on ncurses (among other things).
.P
Searching for packages that actually depend on logrotate:
.nf
	$ \fIqdepends -qQ logrotate\fR
	sys-apps/portage-2.3.18: !<app-admin/logrotate-3.8.0
.fi
The \fB-q\fR flag suppresses the entire list of dependencies here, so we
can quickly see what specific dependency is expressed here.  If
the above had used \fI<logrotate-3\fR, for example, no matches would be
returned.  Not using any specifiers allows to reveal in what way a
dependency is expressed.
.P
Producing a list of packages that need rebuilding after a Perl upgrade:
.nf
	$ \fIqdepends -Qqq -F '%[CATEGORY]%[PN]%[SLOT]' ^perl:0/5.30\fR
	dev-vcs/git:0
	perl-core/File-Temp:0
	dev-perl/Text-CSV_XS:0
	...
.fi
With the double \fB-q\fR flag, just the matching packages are returned,
allowing to use the list for instance as arguments to \fBemerge\fR(1).
The custom formatter in this case ensures all selected packages will be
the best matching for the currently installed SLOT.  Finally, the use of
the antislot matching (\fI^\fR) is returning all packages that depend on
perl in SLOT \fB0\fR, but in a different SUBSLOT than \fB5.30\fR,
effectively requesting the list of packages that need rebuilding to use
the newly installed version of Perl.
.SH "REPORTING BUGS"
Please report bugs via http://bugs.gentoo.org/
.br
Product: Gentoo Linux; Component: Current packages
.SH AUTHORS
.nf
Ned Ludd <solar@gentoo.org>
Mike Frysinger <vapier@gentoo.org>
Fabian Groffen <grobian@gentoo.org>
.fi
.SH "SEE ALSO"
.BR q (1),
.BR qatom (1),
.BR qcheck (1),
.BR qfile (1),
.BR qgrep (1),
.BR qkeyword (1),
.BR qlist (1),
.BR qlop (1),
.BR qmanifest (1),
.BR qmerge (1),
.BR qpkg (1),
.BR qsearch (1),
.BR qsize (1),
.BR qtbz2 (1),
.BR qtegrity (1),
.BR quse (1),
.BR qwhich (1),
.BR qxpak (1)


qdepends.c
/*
 * Copyright 2005-2023 Gentoo Authors
 * Distributed under the terms of the GNU General Public License v2
 *
 * Copyright 2005-2010 Ned Ludd        - <solar@gentoo.org>
 * Copyright 2005-2014 Mike Frysinger  - <vapier@gentoo.org>
 * Copyright 2018-     Fabian Groffen  - <grobian@gentoo.org>
 */

#include "main.h"
#include "applets.h"

#include <ctype.h>
#include <xalloc.h>
#include <assert.h>

#include "atom.h"
#include "dep.h"
#include "set.h"
#include "tree.h"
#include "xarray.h"
#include "xasprintf.h"
#include "xregex.h"

#define QDEPENDS_FLAGS "drpbIQitUF:SR" COMMON_FLAGS
static struct option const qdepends_long_opts[] = {
	{"depend",    no_argument, NULL, 'd'},
	{"rdepend",   no_argument, NULL, 'r'},
	{"pdepend",   no_argument, NULL, 'p'},
	{"bdepend",   no_argument, NULL, 'b'},
	{"idepend",   no_argument, NULL, 'I'},
	{"query",     no_argument, NULL, 'Q'},
	{"installed", no_argument, NULL, 'i'},
	{"tree",      no_argument, NULL, 't'},
	{"use",       no_argument, NULL, 'U'},
	{"format",     a_argument, NULL, 'F'},
	{"pretty",    no_argument, NULL, 'S'},
	{"resolve",   no_argument, NULL, 'R'},
	COMMON_LONG_OPTS
};
static const char * const qdepends_opts_help[] = {
	"Show DEPEND info",
	"Show RDEPEND info",
	"Show PDEPEND info",
	"Show BDEPEND info",
	"Show IDEPEND info",
	"Query reverse deps",
	"Search installed packages using VDB",
	"Search available ebuilds in the tree",
	"Apply profile USE-flags to conditional deps",
	"Print matched atom using given format string",
	"Pretty format specified depend strings",
	"Resolve found dependencies to package versions",
	COMMON_OPTS_HELP
};
#define qdepends_usage(ret) usage(ret, QDEPENDS_FLAGS, qdepends_long_opts, qdepends_opts_help, NULL, lookup_applet_idx("qdepends"))

/* structures / types / etc ... */
struct qdepends_opt_state {
	unsigned int  qmode;
	array_t      *atoms;
	array_t      *deps;
	set          *udeps;
	char         *depend;
	size_t        depend_len;
	const char   *format;
	char          resolve:1;
	tree_ctx     *vdb;
	tree_ctx     *rtree;
};

#define QMODE_DEPEND     (1<<0)
#define QMODE_RDEPEND    (1<<1)
#define QMODE_PDEPEND    (1<<2)
#define QMODE_BDEPEND    (1<<3)
#define QMODE_IDEPEND    (1<<4)
#define QMODE_INSTALLED  (1<<5)
#define QMODE_TREE       (1<<6)
#define QMODE_REVERSE    (1<<7)
#define QMODE_FILTERUSE  (1<<8)

#define QMODE_DEP_FIRST  QMODE_DEPEND
#define QMODE_DEP_LAST   QMODE_IDEPEND

const char *depend_files[] = {  /* keep *DEPEND aligned with above defines */
	/* 0 */ "DEPEND",
	/* 1 */ "RDEPEND",
	/* 2 */ "PDEPEND",
	/* 3 */ "BDEPEND",
	/* 4 */ "IDEPEND",
	/* 5 */ NULL
};

static bool
qdepends_print_depend(FILE *fp, const char *depend)
{
	dep_node *dep_tree;

	dep_tree = dep_grow_tree(depend);
	if (dep_tree == NULL)
		return false;

	if (!quiet)
		fprintf(fp, "DEPEND=\"\n");

	dep_print_tree(fp, dep_tree, 1, NULL, NORM, verbose > 1);

	dep_burn_tree(dep_tree);
	if (!quiet)
		fprintf(fp, "\"\n");

	return true;
}

static int
qdepends_results_cb(tree_pkg_ctx *pkg_ctx, void *priv)
{
	struct qdepends_opt_state *state = priv;
	depend_atom *atom;
	depend_atom *datom;
	depend_atom *fatom;
	bool firstmatch = false;
	char buf[_Q_PATH_MAX];
	const char **dfile;
	size_t i;
	size_t n;
	size_t m;
	int ret = 0;
	dep_node *dep_tree;
	char **d;
	char *depstr;

	/* matrix consists of:
	 * - QMODE_*DEPEND
	 * - QMODE_REVERSE or not
	 *
	 * REVERSE vs forward mode requires a different search strategy,
	 * *DEPEND alters the search somewhat and affects results printing.
	 */

	datom = tree_get_atom(pkg_ctx, false);
	if (datom == NULL)
		return ret;

	if ((state->qmode & QMODE_REVERSE) == 0) {
		/* see if this cat/pkg is requested */
		array_for_each(state->atoms, i, atom) {
			if (atom->blocker != ATOM_BL_NONE ||
					atom->SLOT != NULL ||
					atom->REPO != NULL)
				datom = tree_get_atom(pkg_ctx, true);
			if (atom_compare(datom, atom) == EQUAL) {
				atom = NULL;
				break;
			}
		}

		/* nothing matched */
		if (atom != NULL)
			return ret;

		ret = 1;

		datom = tree_get_atom(pkg_ctx, true);
		printf("%s:", atom_format(state->format, datom));
	}

	xarrayfree_int(state->deps);
	clear_set(state->udeps);

#define get_depstr(X) \
	i == QMODE_DEPEND  ? tree_pkg_meta_get(pkg_ctx, DEPEND)  : \
	i == QMODE_RDEPEND ? tree_pkg_meta_get(pkg_ctx, RDEPEND) : \
	i == QMODE_PDEPEND ? tree_pkg_meta_get(pkg_ctx, PDEPEND) : \
	i == QMODE_BDEPEND ? tree_pkg_meta_get(pkg_ctx, BDEPEND) : \
		                 tree_pkg_meta_get(pkg_ctx, IDEPEND) ;

	dfile = depend_files;
	for (i = QMODE_DEP_FIRST; i <= QMODE_DEP_LAST; i <<= 1, dfile++) {
		if (!(state->qmode & i))
			continue;

		depstr = get_depstr(i);
		if (depstr == NULL)
			continue;
		dep_tree = dep_grow_tree(depstr);
		if (dep_tree == NULL) {
			warn("failed to parse depstring from %s\n", atom_to_string(datom));
			continue;
		}

		/* try and resolve expressions to real package atoms */
		if (state->resolve)
			dep_resolve_tree(dep_tree, state->rtree);

		if (state->qmode & QMODE_TREE &&
			!(state->qmode & QMODE_REVERSE) &&
			verbose)
		{
			/* pull in flags in use if possible */
			tree_cat_ctx *vcat =
				tree_open_cat(state->vdb, pkg_ctx->cat_ctx->name);
			if (vcat != NULL) {
				tree_pkg_ctx *vpkg =
					tree_open_pkg(vcat, pkg_ctx->name);
				if (vpkg != NULL) {
					depstr = get_depstr(i);
					if (depstr != NULL) {
						dep_node *dep_vdb = dep_grow_tree(depstr);
						if (dep_vdb != NULL) {
							dep_flatten_tree(dep_vdb, state->deps);
							dep_burn_tree(dep_vdb);
						} else {
							warn("failed to parse VDB depstring from %s\n",
								 atom_to_string(datom));
						}
					}
					tree_close_pkg(vpkg);
				}
				tree_close_cat(vcat);
			}
		} else {
			if (state->qmode & QMODE_FILTERUSE)
				dep_prune_use(dep_tree, ev_use);
			dep_flatten_tree(dep_tree, state->deps);
		}

		if (verbose) {
			if (state->qmode & QMODE_REVERSE) {
				array_for_each(state->deps, m, atom) {
					array_for_each(state->atoms, n, fatom) {
						if (atom_compare(atom, fatom) == EQUAL) {
							fatom = NULL;
							break;
						}
					}
					if (fatom == NULL) {
						atom = NULL;
						break;
					}
				}
				if (atom == NULL) {
					ret = 1;

					if (!firstmatch) {
						datom = tree_get_atom(pkg_ctx, true);
						printf("%s:", atom_format(state->format, datom));
					}
					firstmatch = true;

					printf("\n%s=\"\n", *dfile);
					dep_print_tree(stdout, dep_tree, 1, state->atoms,
							RED, verbose > 1);
					printf("\"");
				}
			} else {
				printf("\n%s=\"\n", *dfile);
				dep_print_tree(stdout, dep_tree, 1, state->deps,
						GREEN, verbose > 1);
				printf("\"");
			}
		} else {
			if (state->qmode & QMODE_REVERSE) {
				array_for_each(state->deps, m, atom) {
					array_for_each(state->atoms, n, fatom) {
						if (atom_compare(atom, fatom) == EQUAL) {
							fatom = NULL;
							break;
						}
					}
					if (fatom == NULL) {
						ret = 1;

						if (!firstmatch) {
							datom = tree_get_atom(pkg_ctx, true);
							printf("%s%s", atom_format(state->format, datom),
									quiet < 2 ? ":" : "");
						}
						firstmatch = true;

						snprintf(buf, sizeof(buf), "%s%s%s",
								RED, atom_to_string(atom), NORM);
						if (quiet < 2)
							state->udeps = add_set_unique(buf,
														  state->udeps, NULL);
					} else if (!quiet) {
						state->udeps = add_set_unique(atom_to_string(atom),
													  state->udeps, NULL);
					}
				}
			} else {
				array_for_each(state->deps, m, atom)
					state->udeps = add_set_unique(atom_to_string(atom),
												  state->udeps, NULL);
			}
		}

		xarrayfree_int(state->deps);
		dep_burn_tree(dep_tree);
	}
	if (verbose && ret == 1)
		printf("\n");

#undef get_depstr

	if (!verbose) {
		if ((state->qmode & QMODE_REVERSE) == 0 || ret == 1) {
			for (n = list_set(state->udeps, &d); n > 0; n--)
				printf(" %s", d[n -1]);
			free(d);
			printf("\n");
		}
	}

	return ret;
}

int qdepends_main(int argc, char **argv)
{
	depend_atom *atom;
	DECLARE_ARRAY(atoms);
	DECLARE_ARRAY(deps);
	struct qdepends_opt_state state = {
		.atoms   = atoms,
		.deps    = deps,
		.udeps   = create_set(),
		.qmode   = 0,
		.format  = "%[CATEGORY]%[PF]",
		.resolve = false,
		.vdb     = NULL,
		.rtree   = NULL,
	};
	size_t i;
	int ret;
	bool do_pretty = false;

	if (quiet)
		state.format = "%[CATEGORY]%[PN]";

	while ((ret = GETOPT_LONG(QDEPENDS, qdepends, "")) != -1) {
		switch (ret) {
		COMMON_GETOPTS_CASES(qdepends)

		case 'd': state.qmode |= QMODE_DEPEND;    break;
		case 'r': state.qmode |= QMODE_RDEPEND;   break;
		case 'p': state.qmode |= QMODE_PDEPEND;   break;
		case 'b': state.qmode |= QMODE_BDEPEND;   break;
		case 'I': state.qmode |= QMODE_IDEPEND;   break;
		case 'Q': state.qmode |= QMODE_REVERSE;   break;
		case 'i': state.qmode |= QMODE_INSTALLED; break;
		case 't': state.qmode |= QMODE_TREE;      break;
		case 'U': state.qmode |= QMODE_FILTERUSE; break;
		case 'S': do_pretty = true;               break;
		case 'R': state.resolve = true;           break;
		case 'F': state.format = optarg;          break;
		}
	}

	if ((state.qmode & ~(QMODE_REVERSE | QMODE_INSTALLED | QMODE_TREE)) == 0) {
		/* default mode of operation: -drpb (also for just -Q) */
		state.qmode |= QMODE_DEPEND  |
					   QMODE_RDEPEND |
					   QMODE_PDEPEND |
					   QMODE_BDEPEND |
					   QMODE_IDEPEND ;
	}

	/* default to installed packages */
	if (!(state.qmode & QMODE_INSTALLED) && !(state.qmode & QMODE_TREE))
		state.qmode |= QMODE_INSTALLED;

	/* don't allow both installed and from tree */
	if (state.qmode & QMODE_INSTALLED && state.qmode & QMODE_TREE) {
		warn("-i and -t cannot be used together, dropping -i");
		state.qmode &= ~QMODE_INSTALLED;
	}

	if ((argc == optind) && !do_pretty) {
		free_set(state.udeps);
		qdepends_usage(EXIT_FAILURE);
	}

	if (do_pretty) {
		ret = EXIT_SUCCESS;
		while (optind < argc) {
			if (!qdepends_print_depend(stdout, argv[optind++])) {
				ret = EXIT_FAILURE;
				break;
			}
			if (optind < argc)
				fprintf(stdout, "\n");
		}
		free_set(state.udeps);
		return ret;
	}

	argc -= optind;
	argv += optind;

	for (i = 0; i < (size_t)argc; ++i) {
		atom = atom_explode(argv[i]);
		if (!atom)
			warn("invalid atom: %s", argv[i]);
		else
			xarraypush_ptr(atoms, atom);
	}

	if (state.qmode & QMODE_INSTALLED || verbose)
		state.vdb = tree_open_vdb(portroot, portvdb);
	ret = 0;
	if (state.qmode & QMODE_TREE) {
		char *overlay;
		size_t n;
		tree_ctx *t;

		array_for_each(overlays, n, overlay) {
			t = tree_open(portroot, overlay);
			if (t != NULL) {
				if (state.resolve)
					state.rtree = tree_open(portroot, overlay);
				if (!(state.qmode & QMODE_REVERSE) && array_cnt(atoms) > 0) {
					array_for_each(atoms, i, atom) {
						ret |= tree_foreach_pkg_sorted(t,
								qdepends_results_cb, &state, atom);
					}
				} else {
					ret |= tree_foreach_pkg_sorted(t,
							qdepends_results_cb, &state, NULL);
				}
				tree_close(t);
				if (state.rtree)
					tree_close(state.rtree);
			}
		}
	} else {  /* INSTALLED */
		if (state.resolve)
			state.rtree = tree_open_vdb(portroot, portvdb);
		if (!(state.qmode & QMODE_REVERSE) && array_cnt(atoms) > 0) {
			array_for_each(atoms, i, atom) {
				ret |= tree_foreach_pkg_fast(state.vdb,
						qdepends_results_cb, &state, atom);
			}
		} else {
			ret |= tree_foreach_pkg_fast(state.vdb,
					qdepends_results_cb, &state, NULL);
		}
		if (state.rtree)
			tree_close(state.rtree);
	}

	if (state.vdb != NULL)
		tree_close(state.vdb);
	if (state.depend != NULL)
		free(state.depend);

	array_for_each(atoms, i, atom)
		atom_implode(atom);
	xarrayfree_int(atoms);
	xarrayfree_int(state.deps);
	free_set(state.udeps);

	if (!ret)
		warn("no matches found for your query");
	return ret ? EXIT_SUCCESS : EXIT_FAILURE;
}
